File: bindings/matlab/+pqc/kem.m
% libpqc-dyber - Post-Quantum Cryptography Library
% Copyright (c) 2024-2026 Dyber, Inc.
% SPDX-License-Identifier: Apache-2.0 OR MIT
%
% KEM (Key Encapsulation Mechanism) class.

classdef kem < handle
    properties (SetAccess = private)
        Algorithm  char
        Handle     uint64
    end

    properties (Dependent)
        PublicKeySize
        SecretKeySize
        CiphertextSize
        SharedSecretSize
        SecurityLevel
    end

    methods
        function obj = kem(algorithm)
            %KEM Create a KEM context for the specified algorithm.
            %   k = pqc.kem('ML-KEM-768')
            arguments
                algorithm char
            end
            obj.Algorithm = algorithm;
            obj.Handle = pqc_mex('kem_new', algorithm);
            if obj.Handle == 0
                error('pqc:unsupported', 'Unsupported KEM algorithm: %s', algorithm);
            end
        end

        function delete(obj)
            if obj.Handle ~= 0
                pqc_mex('kem_free', obj.Handle);
                obj.Handle = 0;
            end
        end

        function sz = get.PublicKeySize(obj)
            sz = pqc_mex('kem_public_key_size', obj.Handle);
        end

        function sz = get.SecretKeySize(obj)
            sz = pqc_mex('kem_secret_key_size', obj.Handle);
        end

        function sz = get.CiphertextSize(obj)
            sz = pqc_mex('kem_ciphertext_size', obj.Handle);
        end

        function sz = get.SharedSecretSize(obj)
            sz = pqc_mex('kem_shared_secret_size', obj.Handle);
        end

        function lvl = get.SecurityLevel(obj)
            lvl = pqc_mex('kem_security_level', obj.Handle);
        end

        function [pk, sk] = keygen(obj)
            %KEYGEN Generate a keypair.
            %   [pk, sk] = k.keygen()
            %   pk and sk are uint8 column vectors.
            [pk, sk] = pqc_mex('kem_keygen', obj.Handle);
        end

        function [ct, ss] = encaps(obj, pk)
            %ENCAPS Encapsulate a shared secret with a public key.
            %   [ct, ss] = k.encaps(pk)
            arguments
                obj
                pk uint8
            end
            [ct, ss] = pqc_mex('kem_encaps', obj.Handle, pk);
        end

        function ss = decaps(obj, ct, sk)
            %DECAPS Decapsulate a shared secret from ciphertext using secret key.
            %   ss = k.decaps(ct, sk)
            arguments
                obj
                ct uint8
                sk uint8
            end
            ss = pqc_mex('kem_decaps', obj.Handle, ct, sk);
        end
    end
end

File: bindings/rust/src/kem.rs
// libpqc-dyber - Post-Quantum Cryptography Library
// Copyright (c) 2024-2026 Dyber, Inc.
// SPDX-License-Identifier: Apache-2.0 OR MIT
//
//! Safe wrapper around the libpqc KEM API.

use crate::error::{self, Error, Result};
use crate::ffi;
use std::ffi::CString;
use std::ptr;

/// A Key Encapsulation Mechanism context.
///
/// Wraps a `PQC_KEM` handle and provides safe key generation,
/// encapsulation, and decapsulation operations.
///
/// # Example
///
/// ```no_run
/// use pqc_dyber::Kem;
///
/// let kem = Kem::new("ML-KEM-768").unwrap();
/// let (pk, sk) = kem.keygen().unwrap();
/// let (ct, ss_enc) = kem.encaps(&pk).unwrap();
/// let ss_dec = kem.decaps(&ct, &sk).unwrap();
/// assert_eq!(ss_enc, ss_dec);
/// ```
pub struct Kem {
    ctx: *mut ffi::PqcKem,
}

// SAFETY: The underlying C context is internally synchronized and does not
// hold thread-local state; it is safe to move across threads.
unsafe impl Send for Kem {}
unsafe impl Sync for Kem {}

impl Kem {
    /// Create a new KEM context for the given algorithm name.
    ///
    /// Returns [`Error::NotSupported`] if the algorithm is unknown or disabled.
    pub fn new(algorithm: &str) -> Result<Self> {
        let c_alg =
            CString::new(algorithm).map_err(|_| Error::InvalidArgument)?;
        let ctx = unsafe { ffi::pqc_kem_new(c_alg.as_ptr()) };
        if ctx.is_null() {
            Err(Error::NotSupported)
        } else {
            Ok(Kem { ctx })
        }
    }

    /// The algorithm name string.
    pub fn algorithm(&self) -> &str {
        unsafe {
            let ptr = ffi::pqc_kem_algorithm(self.ctx);
            std::ffi::CStr::from_ptr(ptr)
                .to_str()
                .unwrap_or("unknown")
        }
    }

    /// Size of a public key in bytes.
    pub fn public_key_size(&self) -> usize {
        unsafe { ffi::pqc_kem_public_key_size(self.ctx) }
    }

    /// Size of a secret key in bytes.
    pub fn secret_key_size(&self) -> usize {
        unsafe { ffi::pqc_kem_secret_key_size(self.ctx) }
    }

    /// Size of a ciphertext in bytes.
    pub fn ciphertext_size(&self) -> usize {
        unsafe { ffi::pqc_kem_ciphertext_size(self.ctx) }
    }

    /// Size of the shared secret in bytes.
    pub fn shared_secret_size(&self) -> usize {
        unsafe { ffi::pqc_kem_shared_secret_size(self.ctx) }
    }

    /// NIST security level (1-5).
    pub fn security_level(&self) -> i32 {
        unsafe { ffi::pqc_kem_security_level(self.ctx) }
    }

    /// Generate a keypair.
    ///
    /// Returns `(public_key, secret_key)`.
    pub fn keygen(&self) -> Result<(Vec<u8>, Vec<u8>)> {
        let pk_len = self.public_key_size();
        let sk_len = self.secret_key_size();
        let mut pk = vec![0u8; pk_len];
        let mut sk = vec![0u8; sk_len];

        let status = unsafe {
            ffi::pqc_kem_keygen(self.ctx, pk.as_mut_ptr(), sk.as_mut_ptr())
        };
        error::check_status(status)?;
        Ok((pk, sk))
    }

    /// Encapsulate a shared secret using a public key.
    ///
    /// Returns `(ciphertext, shared_secret)`.
    pub fn encaps(&self, pk: &[u8]) -> Result<(Vec<u8>, Vec<u8>)> {
        if pk.len() != self.public_key_size() {
            return Err(Error::InvalidArgument);
        }

        let ct_len = self.ciphertext_size();
        let ss_len = self.shared_secret_size();
        let mut ct = vec![0u8; ct_len];
        let mut ss = vec![0u8; ss_len];

        let status = unsafe {
            ffi::pqc_kem_encaps(
                self.ctx,
                ct.as_mut_ptr(),
                ss.as_mut_ptr(),
                pk.as_ptr(),
            )
        };
        error::check_status(status)?;
        Ok((ct, ss))
    }

    /// Decapsulate a ciphertext using a secret key to recover the shared secret.
    pub fn decaps(&self, ct: &[u8], sk: &[u8]) -> Result<Vec<u8>> {
        if ct.len() != self.ciphertext_size() {
            return Err(Error::InvalidArgument);
        }
        if sk.len() != self.secret_key_size() {
            return Err(Error::InvalidArgument);
        }

        let ss_len = self.shared_secret_size();
        let mut ss = vec![0u8; ss_len];

        let status = unsafe {
            ffi::pqc_kem_decaps(
                self.ctx,
                ss.as_mut_ptr(),
                ct.as_ptr(),
                sk.as_ptr(),
            )
        };
        error::check_status(status)?;
        Ok(ss)
    }

    /// Return the raw FFI pointer. Use with caution.
    pub fn as_ptr(&self) -> *const ffi::PqcKem {
        self.ctx
    }
}

impl Drop for Kem {
    fn drop(&mut self) {
        if !self.ctx.is_null() {
            unsafe { ffi::pqc_kem_free(self.ctx) };
            self.ctx = ptr::null_mut();
        }
    }
}
